crates/noirc_frontend/src/monomorphization/mod.rs

```diff
@@ -953,27 +1000,79 @@ impl<'interner> Monomorphizer<'interner> {
             Param(pattern, typ, noirc_abi::AbiVisibility::Private)
         }));
 
-        let parameters = self.parameters(parameters);
-        let body = self.expr(lambda.body);
+        let converted_parameters = self.parameters(parameters);
 
         let id = self.next_function_id();
-        let return_type = ret_type.clone();
         let name = lambda_name.to_owned();
-        let unconstrained = false;
+        let return_type = ret_type.clone();
 
-        let function = ast::Function { id, name, parameters, body, return_type, unconstrained };
-        self.push_function(id, function);
+        let env_local_id = self.next_local_id();
+        let env_name = "env";
+        let env_tuple = ast::Expression::Tuple(vecmap(&lambda.captures, |capture| {
+            match capture.transitive_capture_index {
+                Some(field_index) => match self.lambda_envs_stack.last() {
+                    Some(lambda_ctx) => ast::Expression::ExtractTupleField(
+                        lambda_ctx.env_ident.clone(),
+                        field_index,
+                    ),
+                    None => unreachable!(
+                        "Expected to find a parent closure environment, but found none"
+                    ),
+                },
+                None => {
+                    let ident = self.local_ident(&capture.ident).unwrap();
+                    ast::Expression::Ident(ident)
+                }
+            }
+        }));
+        let env_typ = ast::type_of(&env_tuple);
```

Comment:
If we stored the env type on the lambda from type checking we could get rid of this type_of call

Reply:
we do store it, now with passing `expr` it seems we can access it